### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by Shestak2039**
Product Service: https://dxrqxeehg8.execute-api.eu-west-1.amazonaws.com/dev/
FE APP: https://d3h30xy12bqq9w.cloudfront.net/
FE PR: https://github.com/Shestak2039/nodejs-aws-fe/pull/3

GET all products: https://dxrqxeehg8.execute-api.eu-west-1.amazonaws.com/dev/products
GET product by id: https://dxrqxeehg8.execute-api.eu-west-1.amazonaws.com/dev/product/2d733218-5490-467a-b00c-1a30ffc68d94
POST create product: https://dxrqxeehg8.execute-api.eu-west-1.amazonaws.com/dev/product
curl --location --request POST 'https://dxrqxeehg8.execute-api.eu-west-1.amazonaws.com/dev/product' \
--header 'Content-Type: application/json' \
--data-raw '{
    "title": "lol",
    "description": "lol",
    "price": 1000,
    "imageUrl": "https://webneel.com/daily/sites/default/files/images/daily/01-2015/3-unusual-product-design.jpg",
    "count": 1
}'

(Done) 4.1. Write SQL script to fill tables with test examples. Store it in your GIT repository. Execute it for your DB to fill data.
(Done) 4.2. Setup DB connection  and integrate getProduct and getProducts lambdas with DB
(Done) 4.3 Add lambda for create product

(Done) Integrate FE app 'Add Product Form' with new POST /product endpoint

Additional tasks:
(Done)+1 - POST/products lambda functions returns error 400 status code if product data is invalid (Validation through Gateway)
(Done)+1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
(Done)+1 - All lambdas do console.log for each incoming requests and their arguments
(Done)+1 - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa)

### CODE REVIEW

File: product-service/product-service/src/models/product.model.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 export interface ProductModel {
-    id: string;
+    id?: string;
```
Comment (loginov-rocks) [inline product-service/product-service/src/models/product.model.ts:2]: To have the flexibility you can create something like ProductData structure which doesn't have any data layer related stuff and extend ProductModel from it by just adding id :)

File: product-service/product-service/src/lambda-functions/create-product.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        const createdProduct = await productService.createProduct(JSON.parse(body));
+
+        if (createdProduct) {
+            return {
+                headers,
+                statusCode: 200,
+                body: JSON.stringify(createdProduct)
+            };
+        } else {
+            return {
```
Comment (loginov-rocks) [inline product-service/product-service/src/lambda-functions/create-product.ts:21]: You can also throw from here so the same 500 error will be returned from the catch :)

File: product-service/product-service/src/lambda-functions/create-product.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+                body: JSON.stringify({
+                    errorMessage: 'Something went wrong'
+                })
+            };
+        }
+    } catch (e) {
+        console.log(e);
+
+        return {
+            statusCode: 500,
```
Comment (loginov-rocks) [inline product-service/product-service/src/lambda-functions/create-product.ts:33]: Are we missing `headers` here as well?

File: product-service/product-service/src/models/product.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                text: `
+                    WITH first_insert as (
+                        INSERT INTO products (title, description, price, imageURL)
+                        VALUES ($1, $2, $3, $4)
+                        RETURNING id
+                    )
+                        INSERT INTO stocks (product_id, count)
+                        VALUES
+                        ( (SELECT id FROM first_insert), $5)
+                        RETURNING (SELECT id FROM first_insert);`,
```
Comment (loginov-rocks) [inline product-service/product-service/src/models/product.ts:64]: Cool solution! By the way, if you have a single query it's automatically wrapped with the transaction if I remember correctly.

**Review by loginov-rocks — APPROVED** [review]
Nice work!

File: product-service/product-service/src/models/product.ts (hunk 4 above)
Comment (Shestak2039) [reply product-service/product-service/src/models/product.ts:64]: thanks

File: product-service/product-service/src/lambda-functions/create-product.ts (hunk 3 above)
Comment (Shestak2039) [reply product-service/product-service/src/lambda-functions/create-product.ts:33]: thanks

File: product-service/product-service/src/models/product.model.ts (hunk 1 above)
Comment (Shestak2039) [reply product-service/product-service/src/models/product.model.ts:2]: thanks

**Pull request merged**